driver-core/src/test/unit/com/mongodb/internal/connection/InternalStreamConnectionInitializerSpecification.groovy

```diff
@@ -243,6 +256,94 @@ class InternalStreamConnectionInitializerSpecification extends Specification {
+        def initializer = new InternalStreamConnectionInitializer(authenticator, null, [])
+
+        when:
+        enqueueSpeculativeAuthenticationResponsesForScramSha256()
+        def description = initializer.initialize(internalConnectionForSpeculativeAuthentication)
+
+        then:
+        description
+        1 * authenticator.authenticate(internalConnectionForSpeculativeAuthentication, _)
+        1 * authenticator.createSpeculativeAuthenticateCommand(internalConnectionForSpeculativeAuthentication)
+    }
+
+    @IgnoreIf({ !serverVersionAtLeast(4, 3) })
+    def 'should speculatively authenticate with SCRAM-SHA-1 authenticator'() {
+        given:
+        def credential = new MongoCredentialWithCache(createScramSha1Credential('user', 'database', 'pencil' as char[]))
+        def authenticator = Spy(ScramShaAuthenticator, constructorArgs: [credential, { 'fyko+d2lbbFgONRv9qkxdawL' }, { 'pencil' }])
+        def initializer = new InternalStreamConnectionInitializer(authenticator, null, [])
+
+        when:
+        enqueueSpeculativeAuthenticationResponsesForScramSha1()
+        def description = initializer.initialize(internalConnectionForSpeculativeAuthentication)
+
+        then:
+        description
+        1 * authenticator.authenticate(internalConnectionForSpeculativeAuthentication, _)
+        1 * authenticator.createSpeculativeAuthenticateCommand(internalConnectionForSpeculativeAuthentication)
+    }
+
+    @IgnoreIf({ !serverVersionAtLeast(4, 3) })
+    def 'should speculatively authenticate with X509 authenticator'() {
+        given:
+        def credential = new MongoCredentialWithCache(createMongoX509Credential())
+        def authenticator = Spy(X509Authenticator, constructorArgs: [credential])
+        def initializer = new InternalStreamConnectionInitializer(authenticator, null, [])
+
+        when:
+        enqueueSpeculativeAuthenticationResponsesForX509()
+        def description = initializer.initialize(internalConnectionForSpeculativeAuthentication)
+
+        then:
+        description
+        1 * authenticator.authenticate(internalConnectionForSpeculativeAuthentication, _)
+        1 * authenticator.createSpeculativeAuthenticateCommand(internalConnectionForSpeculativeAuthentication)
+    }
+
+    @IgnoreIf({ !serverVersionAtLeast(4, 3) })
+    def 'should not speculatively authenticate with Plain authenticator'() {
+        given:
+        def credential = new MongoCredentialWithCache(createPlainCredential('user', 'database', 'pencil' as char[]))
+        def authenticator = Spy(PlainAuthenticator, constructorArgs: [credential])
+        def initializer = new InternalStreamConnectionInitializer(authenticator, null, [])
+
+        when:
+        enqueueSpeculativeAuthenticationResponsesForPlain()
+        def description = initializer.initialize(internalConnectionForSpeculativeAuthentication)
+
+        then:
```

Comment:
As above, this should be asserting the contents of the ismaster command (and that it doesn't contain a speculativeAuthenticate field